# Yibby Gonzalez 


# Árbol de decisión

Según el siguiente árbol de decisión, analice y responda:

![Obama-Clinton decision tree](https://github.com/justmarkham/DAT8/raw/226791169b1cc6df8e8845c12e34e748d5ffaa85/notebooks/images/obama_clinton_tree.jpg)

## 1.1 ¿Qué son las observaciones? ¿Cuántas observaciones hay?


In [ ]:
383 + 70 + 704 + 89 + 185+ 36+ 182+79+52+25+201+83+48+13+56+35

2241

Las observaciones son los condados que participaron en las elecciones presidenciales de estados unidos, los cuales votaron por Hillary Clinton o Barack Obama. Hay 2241 observaciones. 

## 1.2 ¿Cuál es la variable de respuesta?


Es el candidato que gana en el condado. 

## 1.3 ¿Cuáles son las características?¿Cuál es la característica más predictiva?


La característica más predictiva es la proporción de la población del condado que es Afroamericana. Seguido de esta se tiene características como tasa de graduados de secundaria en el condado, la ubicación geográfica del condado, la tasa de pobreza medida en ingresos del condado, la densidad de la población y el resultado del condado de las elecciones anteriores entre Bush y Kerry. 

## 1.4 ¿Cuál es la predicción de clase para el siguiente condado con los siguientes datos: 15% de afroamericanos, 90% de tasa de graduación en la escuela secundaria, ubicado en el Sur, alta pobreza, alta densidad de población?


La predicción es que Obama gana. 

## 1.5 ¿Cuál es el chance predicha para ese mismo condado anteriormente descrito?

In [ ]:
185/36

5.138888888888889

El chance predicho es que la probabilidad de que gane Obama es cinco veces la probabilidad de que gane Clinton. 

## SVM


Para esta tarea consideramos un conjunto de observaciones sobre una serie de variedades de vino tinto y blanco que implican sus propiedades químicas y su clasificación por parte de los catadores. La industria del vino ha experimentado un gran crecimiento en los últimos tiempos debido al aumento del consumo social. El precio del vino depende de un concepto bastante abstracto de apreciación del vino por parte de los catadores, cuya opinión puede tener un alto grado de variabilidad. El precio del vino depende en cierta medida de este factor tan volátil. Otro factor clave en la certificación y evaluación de la calidad del vino son las pruebas fisicoquímicas, que se realizan en laboratorio y tienen en cuenta factores como la acidez, el nivel de pH, la presencia de azúcar y otras propiedades químicas. Para el mercado del vino, sería interesante que la calidad humana de la cata pudiera relacionarse con las propiedades químicas del vino para que el proceso de certificación y evaluación de la calidad estuviera más controlado.

Se dispone de dos conjuntos de datos, uno de los cuales se refiere al vino tinto y cuenta con 1.599 variedades diferentes, y el otro, al vino blanco, con 4.898 variedades. Todos los vinos se producen en una zona concreta de Portugal. Se recogen datos sobre 12 propiedades diferentes de los vinos, una de las cuales es la calidad, basada en datos sensoriales, y el resto son propiedades químicas de los vinos, como la densidad, la acidez, el contenido de alcohol, etc. Todas las propiedades químicas de los vinos son variables continuas. La calidad es una variable ordinal con una clasificación posible de 1 (peor) a 10 (mejor). Cada variedad de vino es catada por tres catadores independientes y la clasificación final asignada es la mediana de la clasificación dada por los catadores.

Se espera que un modelo predictivo desarrollado a partir de estos datos sirva de orientación a los viñedos en cuanto a la calidad y el precio que se espera obtener de sus productos sin depender en gran medida de la volatilidad de los catadores.


In [1]:
import pandas as pd
import numpy as np

In [2]:
data_r = pd.read_csv('https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/Wine_data_red.csv')
data_w = pd.read_csv('https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/Wine_data_white.csv')

In [9]:
data = data_w.assign(type = 'white')

data = data.append(data_r.assign(type = 'red'), ignore_index=True)
data.sample(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
5337            7.0              0.62         0.18             1.5      0.062   
2441            7.2              0.20         0.28             1.6      0.028   
5706            7.4              0.53         0.12             1.9      0.165   
5130            8.5              0.37         0.20             2.8      0.090   
2412            6.3              0.38         0.17             8.8      0.080   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
5337                  7.0                  50.0  0.99510  3.08       0.60   
2441                 13.0                 168.0  0.99203  3.17       1.06   
5706                  4.0                  12.0  0.99702  3.26       0.86   
5130                 18.0                  58.0  0.99800  3.34       0.70   
2412                 50.0                 212.0  0.99803  3.47       0.66   

      alcohol  quality   type  
5337      9.3        5    red  
2441     11.5        6  white  
5706      9.2        5    red  
5130      9.6        6    red  
2412      9.4        4  white

In [4]:
data.quality.value_counts()

6    2836
5    2138
7    1079
4     216
8     193
3      30
9       5
Name: quality, dtype: int64

In [10]:
data.type.value_counts()

white    4898
red      1599
Name: type, dtype: int64

# Ejercicio 2.1

Mostrar la tabla de frecuencias de la calidad por tipo de vino.

In [7]:
!pip install dfply
from dfply import *

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [11]:
df = (data >>
  select(X.type,X.quality) >>
  group_by(X.quality,X.type) >>
  summarize(freq = X.quality.count())
)
df

type  quality  freq
0     red        3    10
1   white        3    20
2     red        4    53
3   white        4   163
4     red        5   681
5   white        5  1457
6     red        6   638
7   white        6  2198
8     red        7   199
9   white        7   880
10    red        8    18
11  white        8   175
12  white        9     5

# Ejercicio 2.2

Teniendo en cuenta todas las variables:

* Crear un objetivo binario para cada tipo de vino, haciendo la transformación necesaria para ingresar la información al modelo.
* Crear una SVM que prediga el tipo de vino
* Evalúe con F1, AUC-ROC y log-loss


In [14]:
Tipo_dummy = pd.get_dummies(data['type'])
Cuali_dummy = pd.get_dummies(data['quality'])

Tipo_dummy = Tipo_dummy.loc[:,['red']] ; # red categoria de referencia 
Cuali_dummy = Cuali_dummy.loc[:,[4, 5, 6, 7, 8, 9]] # 3


In [25]:
y = np.array(Tipo_dummy)
X = pd.concat([data.loc[:,['fixed acidity',	'volatile acidity',	'citric acid',	'residual sugar',	'chlorides',	'free sulfur dioxide',	'total sulfur dioxide',	'density'	,'pH',	'sulphates',	'alcohol']],Cuali_dummy], axis = 1)
X

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0             0.270         0.36            20.7      0.045   
1               6.3             0.300         0.34             1.6      0.049   
2               8.1             0.280         0.40             6.9      0.050   
3               7.2             0.230         0.32             8.5      0.058   
4               7.2             0.230         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
6492            6.2             0.600         0.08             2.0      0.090   
6493            5.9             0.550         0.10             2.2      0.062   
6494            6.3             0.510         0.13             2.3      0.076   
6495            5.9             0.645         0.12             2.0      0.075   
6496            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
6492                 32.0                  44.0  0.99490  3.45       0.58   
6493                 39.0                  51.0  0.99512  3.52       0.76   
6494                 29.0                  40.0  0.99574  3.42       0.75   
6495                 32.0                  44.0  0.99547  3.57       0.71   
6496                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  4  5  6  7  8  9  
0         8.8  0  0  1  0  0  0  
1         9.5  0  0  1  0  0  0  
2        10.1  0  0  1  0  0  0  
3         9.9  0  0  1  0  0  0  
4         9.9  0  0  1  0  0  0  
...       ... .. .. .. .. .. ..  
6492     10.5  0  1  0  0  0  0  
6493     11.2  0  0  1  0  0  0  
6494     11.0  0  0  1  0  0  0  
6495     10.2  0  1  0  0  0  0  
6496     11.0  0  0  1  0  0  0  

[6497 rows x 17 columns]

In [34]:
from sklearn.svm import SVC
from sklearn import metrics
from sklearn.metrics import roc_auc_score

In [32]:
mod1 = SVC()
mod1.fit(X, y)
y_pred = mod1.predict(X)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,


In [35]:
print('F1 score:', metrics.f1_score(y, y_pred))
print('ROC AUC score:', metrics.roc_auc_score(y, y_pred, average=None))
print('Log Loss:', metrics.log_loss(y, y_pred))

F1 score: 0.8619756427604872
ROC AUC score: 0.8899011376802213
Log Loss: 2.1689837056657337


# Ejercicio 2.3

* Probar la SVM utilizando los diferentes kernels ('poly', 'rbf', 'sigmoid')
* Evalúe con F1, AUC-ROC y log-loss


In [36]:
mod2 = SVC(kernel = 'poly')
mod2.fit(X, y)
y_pred = mod2.predict(X)
print('F1 score:', metrics.f1_score(y, y_pred))
print('ROC AUC score:', metrics.roc_auc_score(y, y_pred, average=None))
print('Log Loss:', metrics.log_loss(y, y_pred))

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,


F1 score: 0.8358302122347068
ROC AUC score: 0.8915452466080398
Log Loss: 2.796307384429653


In [37]:
mod2 = SVC(kernel = 'rbf')
mod2.fit(X, y)
y_pred = mod2.predict(X)
print('F1 score:', metrics.f1_score(y, y_pred))
print('ROC AUC score:', metrics.roc_auc_score(y, y_pred, average=None))
print('Log Loss:', metrics.log_loss(y, y_pred))

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,


F1 score: 0.8619756427604872
ROC AUC score: 0.8899011376802213
Log Loss: 2.1689837056657337


In [39]:
mod2 = SVC(kernel = 'sigmoid')
mod2.fit(X, y)
y_pred = mod2.predict(X)
print('F1 score:', metrics.f1_score(y, y_pred))
print('ROC AUC score:', metrics.roc_auc_score(y, y_pred, average=None))
print('Log Loss:', metrics.log_loss(y, y_pred))

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,


F1 score: 0.8263803680981595
ROC AUC score: 0.889146850918206
Log Loss: 3.008957751748889


El mejor SVM es el que cuenta con kernel rbf, ya que la metrica F1 score se acerca mas a uno en comparacio´n con los otros modelos. Ademas este modelo cuenta con mejor Log - loss que los demas. 

# Ejercicio 2.4
* Utilizando el mejor SVM del punto anterior encontrar los parámetros que da el mejor rendimiento teniendo en cuenta los siguientes hiperparámetros
'C': [0.1, 1, 10, 100, 1000], 'gamma': [0.01, 0.001, 0.0001]
* Evalúe cada modelo con F1, AUC-ROC y log-loss

Nota: Para los valores de C y gamma, deben realizar todas las posibles combinaciones entre estas dos variables.

In [43]:
from itertools import combinations
Parametros = [0.1, 1, 10, 100, 1000,0.01, 0.001, 0.0001]
Pares = list(combinations(Parametros, 2))

Tabla = pd.DataFrame(Pares, columns = ['C','gamma'])

grillita = (Tabla >> 
      mask(Tabla['C'] != 0.01, Tabla['C'] != 0.001, Tabla['gamma'] != 1,Tabla['gamma'] != 10,Tabla['gamma'] != 100,Tabla['gamma'] != 1000))
grillita


C   gamma
4      0.1  0.0100
5      0.1  0.0010
6      0.1  0.0001
10     1.0  0.0100
11     1.0  0.0010
12     1.0  0.0001
15    10.0  0.0100
16    10.0  0.0010
17    10.0  0.0001
19   100.0  0.0100
20   100.0  0.0010
21   100.0  0.0001
22  1000.0  0.0100
23  1000.0  0.0010
24  1000.0  0.0001

In [ ]:
F1_mod1 = []
AUCROC_mod1 = []
Logloss_mod1 = []
for i in range(15):
 mod1 = SVC(kernel = 'rbf',C= grillita.iloc[i,0],gamma = grillita.iloc[i,1])
 mod1.fit(X, y)
 y_pred = mod1.predict(X)
 a =  metrics.f1_score(y, y_pred)
 b = metrics.roc_auc_score(y, y_pred)
 c  = metrics.log_loss(y, y_pred)
 F1_mod1.append(a)
 AUCROC_mod1.append(b)
 Logloss_mod1.append(c)

In [47]:
grillita['F1_mod1'] = F1_mod1
grillita['AUCROC_mod1'] = AUCROC_mod1
grillita['Logloss_mod1'] = Logloss_mod1
grillita

C   gamma   F1_mod1  AUCROC_mod1  Logloss_mod1
4      0.1  0.0100  0.860639     0.888012      2.179615
5      0.1  0.0010  0.857239     0.887527      2.248727
6      0.1  0.0001  0.844927     0.878791      2.429476
10     1.0  0.0100  0.897825     0.918439      1.648004
11     1.0  0.0010  0.874119     0.899167      1.993551
12     1.0  0.0001  0.865924     0.892709      2.110506
15    10.0  0.0100  0.965168     0.972975      0.584777
16    10.0  0.0010  0.919675     0.935822      1.313087
17    10.0  0.0001  0.889407     0.910507      1.764957
19   100.0  0.0100  0.996865     0.997084      0.053161
20   100.0  0.0010  0.975087     0.980773      0.419976
21   100.0  0.0001  0.944998     0.955164      0.909060
22  1000.0  0.0100  0.999375     0.999585      0.010632
23  1000.0  0.0010  0.990279     0.992623      0.164801
24  1000.0  0.0001  0.979232     0.984206      0.350866

# Ejercicio 2.5

Compare los resultados con todos los anteriores modelos y eliga ¿Cuál algoritmo con cuáles hiperparámetros es el que tiene mejor rendimiento según la puntuación F1, AUC-ROC y log-loss?

Según las metricas propuestas se encontro que el mejor modelo SVM es el que se ajusta con kernel rbf, con un C de 1000 y un gammas de 0.01. El modleo seleccionado cuenta  con el mayor F1 score y AUC- ROC, ademas tiene el menor Log-loss de todos los modelos propuestos. 